### DISCUSSION TITLE
Failure writing SVG files with IM

### DISCUSSION THREAD

**Discussion by exekutive** [ImageMagick/ImageMagick#7477] [Help]
### ImageMagick version

7.1.1

### Operating system, version and so on

darwin

### Description

Can IM resize SVG files?

I am getting an error.

```
magick logistic-van.svg -resize 50x50 logistic-van-small.svg

(process:6279): GLib-CRITICAL **: 02:39:39.442: g_hash_table_lookup: assertion 'hash_table != NULL' failed
zsh: abort      magick logistic-van.svg -resize 50x50 logistic-van-small.svg

```
![logistic-van](https://github.com/user-attachments/assets/a09fea77-9b28-486a-8b03-e28c7e604cce)

```
magick -version
Version: ImageMagick 7.1.1-35 (Beta) Q16-HDRI aarch64 b0b7b1730:20240623 https://imagemagick.org
Copyright: (C) 1999 ImageMagick Studio LLC
License: https://imagemagick.org/script/license.php
Features: Cipher DPC HDRI Modules OpenCL OpenMP 
Delegates (built-in): autotrace bzlib cairo djvu fftw fontconfig freetype gslib gvc heic jng jp2 jpeg jxl lcms lqr ltdl lzma openexr png ps raw rsvg tiff webp wmf x xml zip zlib zstd
Compiler: gcc (4.2)
```

**Comment by fmw42** [comment]
I get the same message with IM 7.1.1.34 GS 10.03 on Mac OSX Ventura, but the output is created. It is a Ghostscript issue.

`magick MSVG:file.svg -resize 50x50 x.png`

<br>

This works fine with no errors using MSVG

`magick MSVG:file.svg -resize 50x50 y.png`

**Reply by exekutive** [reply]
you are outputting PNG files. I need SVG

**Reply by fmw42** [reply]
Works fine for me, but I still get the GS message.

```
magick file.svg -resize 50x50 x.svg
magick x.svg x.png
```

![x](https://github.com/user-attachments/assets/017969be-503c-4ee2-a15e-471d24871d17)

**Reply by fmw42** [reply]
P.S. I processed in a bash window rather than a zsh window.  Don't know if that makes a difference.

**Reply by exekutive** [reply]
I get an output file but it's zero bytes. Interesting, I'll have to try bash.

**Comment by exekutive** [comment]
it seems I'm unable to write SVG files in any capacity using IM, with ZSH or BASH. I always get the hash table failure.


```
 GLib-CRITICAL **: 12:04:56.525: g_hash_table_lookup: assertion 'hash_table != NULL' failed
Abort trap: 6

```

Ghostscript is an interpreter for Adobe Postscript and PDF files. It can't touch SVG. I believe SVG is supported through the `librsvg` delegate (which I have ver 2.58.1)

**Comment by fmw42** [comment]
To read an SVG file, you need either Inkscape or the RSVG delegate or use IM's internal MSVG renderer.   I am no expert, but to write an SVG you might need potrace.

**Reply by fmw42** [reply]
On my Mac Ventura with Inkscape installed, I get

```
magick file.svg -resize 50x50 x.svg

(<unknown>:1737): glibmm-WARNING **: 12:23:26.505: Failed to wrap object of type 'GtkRecentManager'. Hint: this error is commonly caused by failing to call a library init() function.

But the resulting file is created.  
```

But this works without issue:

`magick MSVG:file.svg -resize 50x50 x.svg`


P.S. My IM is 

```
magick  -version
Version: ImageMagick 7.1.1-34 Q16-HDRI aarch64 22301 https://imagemagick.org
Copyright: (C) 1999 ImageMagick Studio LLC
License: https://imagemagick.org/script/license.php
Features: Cipher DPC HDRI Modules 
Delegates (built-in): bzlib cairo djvu fftw fontconfig freetype gslib heic jbig jng jp2 jpeg jxl lcms ltdl lzma openexr pangocairo png ps raw rsvg tiff webp x xml zlib zstd
Compiler: gcc (4.2)
```

**Reply by exekutive** [reply]
OK so you don't have `potrace` as a delegate either. It doesn't seem to be listed in compile options (https://imagemagick.org/script/advanced-linux-installation.php)

`magick MSVG:file.svg -resize 50x50 x.svg` ran without error, but it produces a garbage file containing only :

```
<?xml version="1.0" standalone="no"?>
<!DOCTYPE svg PUBLIC "-//W3C//DTD SVG 20010904//EN"
  "http://www.w3.org/TR/2001/REC-SVG-20010904/DTD/svg10.dtd">
<svg width="50" height="50">
<g style="fill:black;fill-opacity:1;stroke:none;stroke-width:1;stroke-opacity:0;fill-rule:nonzero;</svg>

```

**Reply by exekutive** [reply]
potrace is for tracing bitmaps. I don't think it's relevant either.
https://potrace.sourceforge.net/

**Comment by exekutive** [comment]
I installed inkscape for CLI, and still get zero byte output

```
magick boat.svg -resize 50x50 x.svg     
2024-07-13 12:50:44.817 inkscape[28750:1744512] WARNING: Secure coding is not enabled for restorable state! Enable secure coding by implementing NSApplicationDelegate.applicationSupportsSecureRestorableState: and returning YES.

(process:28749): GLib-CRITICAL **: 12:50:44.832: g_hash_table_lookup: assertion 'hash_table != NULL' failed
zsh: abort      magick boat.svg -resize 50x50 x.svg

```

**Reply by fmw42** [reply]
You have a beta version of IM. There may be developments going on that might cause these issues.  Try installing a released version and see if that helps.  Why did you install a beta?

**Reply by exekutive** [reply]
It wasn't my intention. I just used
`git clone https://github.com/ImageMagick/ImageMagick.git`

**Reply by fmw42** [reply]
You should not get it there.  Download from 

https://imagemagick.org/script/download.php

or get it from Homebrew or MacPorts

**Reply by exekutive** [reply]
the homebrew one is total garbage. That's why I compile it myself. I can't use macports due to homebrew conflict.

**Reply by fmw42** [reply]
If you want source, then get it at 

https://imagemagick.org/archive/

but you still need to install all the delegates that you want.  You can do that using MacPorts.  That is what I do -- install delegates from MacPorts then install IM from source.

**Reply by exekutive** [reply]
I would have to switch my entire tool ecosystem from HB to MP, which would be a multi-week project. I don't have sufficient reason to undertake such a project right now.
I have IM nicely tuned to work the way I want with a compile script right now.
For example I see that you do not have multi-threading or hardware acceleration enabled for your macports IM

**Reply by fmw42** [reply]
Don't expect IM to work successfully from a beta!  Get a release from the GIT then, if you don't like my other solutions.

**Comment by snibgo** [comment]
> Can IM resize SVG files?

No. IM doesn't resize files. In particular, it can't parse a vector SVG file and adjust internal values to have the effect of resizing the vector image.

Ghostscript is not used for reading or writing SVG.

After successfully reading an SVG, IM has a raster image in memory. The result from writing a raster image to SVG should be either an SVG with embedded raster image, or a call to external delegate `potrace` which converts a raster image to a vector SVG. `potrace` is only used if there is a suitable entry in delegates.xml

For example:
```
magick xc: x.svg
```
... makes this SVG:
```
<?xml version="1.0" encoding="UTF-8" standalone="no"?>
<!DOCTYPE svg PUBLIC "-//W3C//DTD SVG 1.1//EN" "http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd">
<svg version="1.1" id="Layer_1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" x="0px" y="0px" width="1px" height="1px" viewBox="0 0 1 1" enable-background="new 0 0 1 1" xml:space="preserve">  <image id="image0" width="1" height="1" x="0" y="0"
    xlink:href="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABAQAAAAA3bvkkAAAAIGNIUk0AAHomAACAhAAA+gAAAIDo
AAB1MAAA6mAAADqYAAAXcJy6UTwAAAACYktHRAAB3YoTpAAAAAd0SU1FB+gHDgMdKKyj4tMAAAAK
SURBVAjXY2gAAACCAIHdQ2r0AAAAJXRFWHRkYXRlOmNyZWF0ZQAyMDI0LTA3LTE0VDAzOjI5OjQw
KzAwOjAw1G0JQgAAACV0RVh0ZGF0ZTptb2RpZnkAMjAyNC0wNy0xNFQwMzoyOTo0MCswMDowMKUw
sf4AAAAodEVYdGRhdGU6dGltZXN0YW1wADIwMjQtMDctMTRUMDM6Mjk6NDArMDA6MDDyJZAhAAAA
AElFTkSuQmCC" />
</svg>
```
Note the embedded raster image.

Tested with IM v7.1.1-20 on Windows 11.

>  GLib-CRITICAL **: 12:50:44.832: g_hash_table_lookup: assertion 'hash_table != NULL' failed

This error is from GLib, which is used by IM's coders/svg.c. GLib has a function `g_hash_table_lookup()`, which is not called directly by IM.

svg.c seems to only call one function in GLib, `g_type_init()`, and only if the version of GLib is before 2.35.0, when that function was deprecated.

**Reply by exekutive** [reply]
Thanks

**Comment by exekutive** [comment]
Can anybody suggest a good command line tool for resizing SVGs?

**Comment by snibgo** [comment]
I'm not sure what you mean by "resizing SVGs". A vector SVG has dimensions in "user units", which can be whatever dimension you want when you rasterize the image.

You can multiply SVG object dimensions (in "user units") like this:
```
inkscape --export-filename=out.svg --actions=select-all;transform-scale:5;fit-canvas-to-selection logistic-van.svg
```

**Reply by exekutive** [reply]
I am a little new to SVGs. I know they are vectors and technically don't have a dimension. I but I believe they still contain some kind of metadata which dictates their size when rendered on a screen.  The web software I encountered (phpBB for example) does not allow you to specify render dimensions and instead seems to use this internal metadata to choose the size. Maybe it is the "user units". Thanks for the inkscape command I'll try that.
